pkg/transport/managed_transport.go

```diff
@@ -135,28 +136,25 @@ func (mt *ManagedTransport) Serve(readCh chan<- routing.Packet, done <-chan stru
 		}
 	}()
 
-	// Redial loop.
-	for {
-		select {
-		case <-mt.done:
-			return
+	r := netutil.NewRetrier(3*time.Second, 5, 2)
+	var err error
 
-		case <-logTicker.C:
-			if mt.logMod() {
-				if err := mt.ls.Record(mt.Entry.ID, mt.LogEntry); err != nil {
-					mt.log.Warnf("Failed to record log entry: %s", err)
-				}
-			} else {
-				// If there has not been any activity, ensure underlying 'write' tp is still up.
-				mt.connMx.Lock()
-				if mt.conn == nil {
-					if err := mt.dial(ctx); err != nil {
-						mt.log.Warnf("failed to redial underlying connection: %v", err)
-					}
-				}
-				mt.connMx.Unlock()
-			}
+	// redial underlying connection
+	err = r.Do(func() error {
+		if mt.logMod() {
+			err = mt.ls.Record(mt.Entry.ID, mt.LogEntry)
+			return err
 		}
+		mt.connMx.Lock()
```

Comment:
What do you think of writing in the following way?

```golang
	mt.connMx.Lock()
	defer mt.connMx.Unlock()

 	return mt.dial(ctx)
```

Reply:
Technically avoiding the `defer` statement can increase performance, but only by a minuscule amount. Personally I am okay with both styles.